File: src/main/scala/com/pwootage/oc/js/jsvalue/JSValue.scala
package com.pwootage.oc.js.jsvalue

import java.nio.charset.StandardCharsets
import java.text.NumberFormat
import java.util
import java.util.function.BiConsumer

import com.google.gson._

import scala.collection.JavaConversions._

/**
  * Basically a small json library wrapping gson because I'm too lazy to figure out an external dep
  */
object JSValue {
  val gson = new GsonBuilder().create()

  def fromJSON(json: String): JSValue = {
    val jsonParser = new JsonParser()
    val ele = jsonParser.parse(json)
    JSValue.fromJSON(ele)
  }

  def fromJSON(ele: JsonElement): JSValue = {
    ele match {
      case primitive: JsonPrimitive if ele.isJsonPrimitive =>
        if (primitive.isBoolean) {
          JSBooleanValue(primitive.getAsBoolean)
        } else if (primitive.isNumber) {
          JSDoubleValue(primitive.getAsDouble)
        } else {
          JSStringValue(primitive.getAsString)
        }
      case array: JsonArray if ele.isJsonArray =>
        val res = new Array[JSValue](array.size())
        for (i <- res.indices) {
          res(i) = fromJSON(array.get(i))
        }
        JSArray(res)
      case obj: JsonObject if ele.isJsonObject =>
        val res = new util.HashMap[String, JSValue]()
        for (entry <- obj.entrySet().iterator()) {
          res.put(entry.getKey, fromJSON(entry.getValue))
        }
        JSMap(res)
      case nul: JsonNull if ele.isJsonNull =>
        JSNull
      case x =>
        println("Unknown json?", x)
        JSNull
    }
  }

  def fromJava[T](obj: T): JSValue = {
    obj match {
      case x: java.lang.Float => JSDoubleValue(x.doubleValue())
      case x: java.lang.Double => JSDoubleValue(x.doubleValue())
      case x: java.lang.Integer => JSDoubleValue(x.doubleValue())
      case x: java.lang.Long => JSDoubleValue(x.doubleValue())
      case x: java.lang.Boolean => JSBooleanValue(x)
      case x: String => JSStringValue(x)
      case x: util.HashMap[kt, vt] =>
        val res = new util.HashMap[String, JSValue]()
        x.forEach(new BiConsumer[kt, vt] {
          override def accept(k: kt, v: vt): Unit = {
            k match {
              case x: String =>
                val jsVal = fromJava(v)
                res.put(x, jsVal)
              case _ => //Do nothing
            }
          }
        })
        JSMap(res)
      case x: Array[Byte] =>
        //TODO: arraybuffers?
        JSStringValue(new String(x, StandardCharsets.UTF_8))
      case x: Array[AnyRef] =>
        JSArray(x.map(fromJava))
      case x: util.ArrayList[_] =>
        JSArray(x.map(fromJava).toArray)
      case null =>
        JSNull
      case x =>
        JSValue.fromJSON(JSValue.gson.toJson(x))
      //        JSStringValue(x.toString)
    }
  }
}

/**
  * Created by pwootage on 5/14/17.
  */
trait JSValue {
  def property(name: String): JSValue = JSNull

  def arrayVal(index: Int): JSValue = JSNull

  def asString: Option[String] = None

  def asDouble: Option[Double] = None

  def asBoolean: Option[Boolean] = None

  def asArray: Option[Array[JSValue]] = None

  /**
    * Converts this to a simple java type. Note this will return None for non-basic types!
    */
  def asSimpleJava: AnyRef

  def toJSON: String
}

case class JSStringValue(value: String) extends JSValue {
  override def asString = Some(value)

  override def asSimpleJava: String = value

  override def toJSON = JSValue.gson.toJson(value)
}

case class JSDoubleValue(value: Double) extends JSValue {

  override def asDouble = Some(value)

  def asLong = {
    val longV = value.toLong
    if (longV == value) {
      Some(longV)
    } else {
      None
    }
  }

  def asInt = {
    val intV = value.toInt
    if (intV == value) {
      Some(intV)
    } else {
      None
    }
  }

  override def asSimpleJava = asInt.map(new java.lang.Integer(_))
    .orElse(asLong.map(new java.lang.Long(_)))
    .getOrElse(new java.lang.Double(value))

  override def toJSON = value.formatted("%f")
}

case class JSBooleanValue(value: Boolean) extends JSValue {
  override def asBoolean = Some(value)

  override def asSimpleJava: java.lang.Boolean = value

  override def toJSON = if (value) "true" else "false"
}

case class JSArray(value: Array[JSValue]) extends JSValue {
  override def arrayVal(index: Int) = {
    if (index >= 0 && index < value.length) {
      value(index)
    } else {
      JSNull
    }
  }

  override def asArray = Some(value)

  override def asSimpleJava = value.map(_.asSimpleJava)

  override def toJSON = s"[${value.map(_.toJSON).mkString(",")}]"
}

case class JSMap(value: util.HashMap[String, JSValue]) extends JSValue {
  override def property(name: String) = value.getOrDefault(name, JSNull)

  override def asSimpleJava: util.HashMap[String, AnyRef] = {
    val res = new util.HashMap[String, AnyRef]()
    value.entrySet().foreach(v => res.put(v.getKey, v.getValue.asSimpleJava))
    res
  }

  override def toJSON = "{" +
    value.entrySet()
      .map(entry => JSValue.gson.toJson(entry.getKey) + ":" + entry.getValue.toJSON)
      .mkString(",") +
    "}"
}

object JSMap {
  def apply(pairs: (String, JSValue)*): JSMap = {
    val res = new util.HashMap[String, JSValue]()
    for ((key, value)<- pairs) {
      res.put(key, value)
    }
    JSMap(res)
  }
}

case object JSNull extends JSValue {
  override def toJSON = "null"

  override def asSimpleJava = null
}

File: src/main/java/com/pwootage/oc/js/JSValue.kt
package com.pwootage.oc.js

import com.google.gson.*

/**
 * Created by pwootage on 5/14/17.
 */
sealed class JSValue {
  open fun property(name: String): JSValue = JSNull

  open fun arrayVal(index: Int): JSValue = JSNull

  open val asString: String? get() = null

  open val asDouble: Double? get() = null

  open val asBoolean: Boolean? get() = null

  open val asArray: Array<JSValue>? get() = null

  /**
   * Converts this to a simple java type. Note this will return None for non-basic types!
   */
  abstract val asSimpleJava: Any?

  abstract fun toJSON(): String

  companion object {
    val gson = GsonBuilder().create()

    fun fromJSON(json: String): JSValue {
      val jsonParser = JsonParser()
      val ele = jsonParser.parse(json)
      return fromJSON(ele)
    }

    fun fromJSON(ele: JsonElement): JSValue {
      return when (ele) {
        is JsonPrimitive ->
          when {
            ele.isBoolean -> JSBooleanValue(ele.getAsBoolean())
            ele.isNumber -> JSDoubleValue(ele.getAsDouble())
            else -> JSStringValue(ele.getAsString())
          }
        is JsonArray -> {
          val res = Array(ele.size()) {
            fromJSON(ele[it])
          }
          JSArray(res)
        }
        is JsonObject -> {
          val res = HashMap<String, JSValue>()
          for (entry in ele.entrySet()) {
            res[entry.key] = fromJSON(entry.value)
          }
          JSMap(res)
        }
        is JsonNull -> JSNull
        else -> {
          println("Unknown json? $ele")
          JSNull
        }
      }
    }

    fun <T : Any> fromJava(obj: T?): JSValue {
      return when (obj) {
        is Float -> JSDoubleValue(obj.toDouble())
        is Double -> JSDoubleValue(obj.toDouble())
        is Int -> JSDoubleValue(obj.toDouble())
        is Long -> JSDoubleValue(obj.toDouble())
        is Boolean -> JSBooleanValue(obj)
        is String -> JSStringValue(obj)
        is HashMap<*, *> -> {
          val res = HashMap<String, JSValue>()
          obj.forEach {
            val k = it.key as? String
            if (k != null)
              res[k] = fromJava(it.value)
          }
          JSMap(res)
        }
        is ByteArray -> JSByteArray(obj)
        is Array<*> -> JSArray(obj.map(::fromJava).toTypedArray())
        is ArrayList<*> -> JSArray(obj.map(::fromJava).toTypedArray())
        null -> JSNull
        else -> {
          println("Unknown fromJava: ${obj::class.java} $obj")
          fromJSON(gson.toJson(obj))
        }
      }
    }
  }
}

data class JSStringValue(val value: String) : JSValue() {
  override val asString = value

  override val asSimpleJava: String = value

  override fun toJSON(): String = gson.toJson(value)
}

data class JSDoubleValue(val value: Double) : JSValue() {

  override val asDouble = value

  override val asSimpleJava = value

  override fun toJSON() = String.format("%f", value);
}

data class JSBooleanValue(val value: Boolean) : JSValue() {
  override val asBoolean = value

  override val asSimpleJava = value

  override fun toJSON() = if (value) "true" else "false"
}

data class JSArray(val value: Array<JSValue>) : JSValue() {
  override fun arrayVal(index: Int): JSValue {
    return if (index >= 0 && index < value.size) {
      value[index]
    } else {
      JSNull
    }
  }

  override val asArray = value

  override val asSimpleJava = value.map { it.asSimpleJava }.toTypedArray()

  override fun toJSON() = "[${value.joinToString(",") { it.toJSON() }}]"
}

data class JSMap(val value: Map<String, JSValue>) : JSValue() {
  override fun property(name: String) = value[name] ?: JSNull

  override val asSimpleJava: Map<String, Any?>
    get() {
      val res = HashMap<String, Any?>()
      value.forEach {
        res[it.key] = it.value.asSimpleJava
      }
      return res
    }

  override fun toJSON() = "{" +
    value.map { gson.toJson(it.key) + ":" + it.value.toJSON() }.joinToString(",") +
    "}"
}

data class JSByteArray(val value: ByteArray) : JSValue() {
  override val asSimpleJava = value

  override fun toJSON(): String {
    TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
  }
}

object JSNull : JSValue() {
  override fun toJSON() = "null"

  override val asSimpleJava: Nothing? = null
}